!!! FEATURE: Render `ResponseInterface` directly in Fusion Views

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

Neos Adjustments for https://github.com/neos/flow-development-collection/pull/3286

 - adjusts Views to return `StreamInterface` or an `ResponseInterface`
 - adjusts Fusion runtime to return `StreamInterface` or an `ResponseInterface`

The fusion runtime will not accept the controller context from outside anymore, but instead has its own legacy layer.
The legacy layer only works in combination with the new entry point in the fusion runtime:

```php
$httpResponse = $fusionRuntime->renderEntryPathWithContext('root', ['node' => $node]);
```

### Fusion's `FusionView` adjustments:

The current view behaves incorrect. The ViewInterface allows to return `string|ActionResponse|ResponseInterface|StreamInterface|\Stringable`
But we violate this by returning the raw runtime value instead ... which for all we now _could_ be a string but might also be just as annotated anything else (`mixed`) not included in the allowed return types, like raw arrays or non string-able objects or integers.
Also previously when using the `Neos.Fusion:Http.Message` in combination with the view it would actually render the http response as string, so `HTTP/` was part of the content and visible in the dom (This is currently broken since Neos 8 and before).

To streamline the behaviour and fix these specialities, the `FusionView` will return if applicable a `ResponseInterface` and for simple (string) cases a psr `StreamInterface`. Both cases have to be anticipated on call site.
The change is not expected to be super breaking if a `StreamInterface` will be returned as for classes with loose typing this will be converted via `__toString`.

This is also necessary for `Neos.Neos:Plugin` and `Neos.Fusion.Form` to work correctly.

Resolves: #4855

**Upgrade instructions**

Mutating the response of the Fusion Runtime (`$this->getRuntime()->getControllerContext()->getResponse()`) is highly internal and only allowed for Neos special cases as the fusion plugin implementation and fusion forms. This will be removed when we find a better concept internally!

Also other usages of the `getControllerContext` are deprecated like accessing the request the following way:

```php
$this->runtime->getControllerContext()->getRequest();
```

Instead, the new fusion globals https://github.com/neos/neos-development-collection/pull/4425 should be used (which will also guarantee that the request in EEL `${request}` is the same as when fetched from a Fusion object:

```php
$possibleRequest = $this->runtime->fusionGlobals->get('request');
if (!$possibleRequest instanceof ActionRequest) {
    // ... fallback
}
```

If you use the Fusion Runtime directly, which is internal, you should know that using `Runtime::render` as entry point will fail to render `Neos.Neos:Plugin`, `Neos.Fusion.Form` and also lead to weird behaviour when rendering a `Neos.Fusion:Http.Message`.
The Fusion view should be used for interacting with Fusion, but if you must, you can leverage `Runtime::renderEntryPathWithContext` to render a full page with all special cases.

**Review instructions / WHY**

The **WHY** of this PR might not seem obvious as a lot of pieces fall here into place.

- We want to deprecate the controller context https://github.com/neos/flow-development-collection/issues/3152
  - -> But removing usages of the controller context is not easily doable as major parts of the system depend on it, like the fusion runtime
- This pr is part of a larger movement and will make Fusion independent of the controller context
- The controller context in Fusion is currently used for 2 major tasks
  1. accessing everything request related (the request itself, the uribuilder and flashmessages) 
  2. accessing the current to be send response by object reference
- the first (i) is easily replaceable by migrating to use the Fusion global "request"
- the second (ii) is currently used for dark magic in Neos.Neos, we modify the to be send response by adding headers or changing the status code
  - in `Neos.Neos:Plugin` for sub-request [link](https://github.com/neos/neos-development-collection/blob/d1eb6ab3d16bf0981cfff31bc8a9a29d69a42350/Neos.Neos/Classes/Fusion/PluginImplementation.php#L169-L180)
  - in `Neos.Fusion:Form` to redirect on success to another page [link](https://github.com/neos/fusion-form/blob/642f728976bd205205ce76180b4c701a67adb30c/Classes/Runtime/FusionObjects/RuntimeFormImplementation.php#L172-L174)
  - instead of creating a slightly prettier but still hacky api for these usecases (like `$runtime->setHeader()`) we will not change the syntax but deprecate it and internalise it.
  - to still support the usecases the Fusion runtime has to create a mutable object (a stub response) itself where it will extract the changes made to it and apply it to an actual reponse (merge it with the contents).
    - for that to archive we require to have a single entry point in Fusion to render a whole page (`renderEntryPathWithContext`)
    - and it requires this entrypoint to return something 'more' than the contents of the page, but also its headers: a full psr response
      - -> this is partially breaking
- to support this change the `ViewInterface` will be adjusted in Flow to force returning either a `StreamInterface` or an `ResponseInterface`, and also the `setControllerContext` will not be necessary to implement / call any longer on views.


**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
WIP: FEATURE: Fusion Runtime render directly http response

mhsdesign committed:
WIP: Neos Fusion View use Runtime::renderResponse

mhsdesign committed:
WIP: Hacky HttpResponseConstraints

mhsdesign committed:
WIP: Hurray make Runtime independent of ControllerContext

+ working legacy layer for manipulating getResponse and accessing getRequest

mhsdesign committed:
WIP: Migrate PluginImplementation to use `unsafeHttpResponseConstrains`

mhsdesign committed:
WIP: FusionObject to not depend on the controller context

mhsdesign committed:
WIP: Skip non working test :(

mhsdesign commented on Neos.Fusion/Classes/View/FusionView.php:
```diff
@@ -46,7 +47,8 @@ class FusionView extends AbstractView
         'debugMode' => [false, 'Flag to enable debug mode of the Fusion runtime explicitly (overriding the global setting).', 'boolean'],
-        'enableContentCache' => [false, 'Flag to enable content caching inside Fusion (overriding the global setting).', 'boolean']
+        'enableContentCache' => [false, 'Flag to enable content caching inside Fusion (overriding the global setting).', 'boolean'],
+        'renderHttpResponse' => [false, 'Flag to render fusion as http repose for advanced form support and Neos.Fusion:Http.ResponseHead support.', 'boolean'],
```
without this flag a lot of test would fail, when we try to return directly an array or object from the runtime:

> Cannot render %s into http response body.

mhsdesign replied:
The more i think about it, the more i want to invert that flag and make it breaking.
Our tests could easily set that flag to render the mixed output directly.

My reasoning for this is that the current Fusion FusionView behaves incorrect.
The ViewInterface allows to return `string|ActionResponse|ResponseInterface|StreamInterface|\Stringable`
But we violate this by returning the raw runtime value instead, which is `mixed` as it could contain raw arrays or non string-able objects or integers...

By introducing a flag like `returnMixed` or `unsafeRawMode` we empathise that this is not the correct way and it violates the interface.

Ideally people using the FusionView should not notice this change as returning a `ResponseInterface` is okay by definition and must (is) accounted by the controllers.

As a migration people could either use the propsed unsafe rawmode, or call `getContents` on the response to get a string.



EDIT
Moved to https://github.com/neos/neos-development-collection/pull/4899

mhsdesign commented on Neos.Fusion/Tests/Functional/View/FusionViewTest.php:
```diff
@@ -64,6 +65,41 @@ public function fusionViewOutputsVariable()
+    /**
+     * @test
+     */
+    public function fusionViewCanReturnHttpResponseFromHttpMessagePrototype()
```
this is actually a bugfix as the FusionView dindt handle `Neos.Fusion:Http.Message` before. Only the Neos FusionView did.

mhsdesign commented on Neos.Neos/Classes/View/FusionView.php:
```diff
@@ -130,35 +125,6 @@ public function render(): string|ResponseInterface
-     * @return string|ResponseInterface If output is a string with a HTTP preamble a ResponseInterface
-     *                                  otherwise the original output.
-     */
-    protected function parsePotentialRawHttpResponse($output)
-    {
-        if ($this->isRawHttpResponse($output)) {
-            return Message::parseResponse($output);
-        }
-
-        return $output;
-    }
```
this hack could be removed from the Neos FusionView and the view will use the new Runtime::renderResponse which will handle this.

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -103,6 +107,8 @@ class Runtime
      */
     public readonly FusionGlobals $fusionGlobals;
 
+    public readonly HttpResponseConstraints $unsafeHttpResponseConstrains;
```
todo add comment that this, like before with response manipulation of the controller context, will only work consistently for uncached segments. For other cases it might only be used once for the first uncached request.

mhsdesign commented on Neos.Neos/Tests/Unit/Fusion/PluginImplementationTest.php:
```diff
@@ -96,12 +99,12 @@ public function responseHeadersDataProvider(): array
                 'Plugin response key does already exist in parent with same value',
                 ['parent' => ['key' => 'value'], 'plugin' => ['key' => 'value']],
-                ['key' => 'value']
+                ['key' => 'value'] // 'value, value'
```
> I think multiple competing headers like that are a broken use case anyways. Headers by definition can appear multiple times, we can't _really_ know if we should remove the first one and when not. IMHO the test is misleading the result might as well (correctly) be key => [value, value]

mhsdesign replied:
Somehow the test now dont fail anymore ... but thats okay because as said the test are odd.

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -148,15 +154,7 @@ public function __construct(
-     * @deprecated {@see self::getControllerContext()}
-     * @internal
-     */
-    public function setControllerContext(ControllerContext $controllerContext): void
```
this method was only introduced in 9.0 dev thats why we can so easily remove it: https://github.com/neos/neos-development-collection/pull/4425

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -169,23 +167,10 @@ public function setControllerContext(ControllerContext $controllerContext): void
      */
     public function getControllerContext(): ControllerContext
     {
-        if (isset($this->controllerContext)) {
```
this logic was only introduced in 9.0 dev thats why we can so easily change it: https://github.com/neos/neos-development-collection/pull/4425

mhsdesign committed:
TASK: Migrate further fusion objects to `$this->runtime->fusionGlobals->get('request');`

mhsdesign committed:
Revert introduction of `HttpResponseConstraints` and api in runtime to allow setting headers dynamically

Instead, the legacy layer

> $this->runtime->getControllerContext()->getResponse();

should be continued to be used by fusion forms and fusion plugin impl.

mhsdesign committed:
TASK: Migrate further fusion objects to `$this->runtime->fusionGlobals->get('request');`

mhsdesign committed:
TASK: Document legacy Runtime::getControllerContext

mhsdesign committed:
TASK: Extract controller context legacy layer into withSimulatedLegacyControllerContext

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -298,6 +264,43 @@ public function getLastEvaluationStatus()
+        }
+        $this->pushContextArray($contextArray);
+
+        return $this->witSimulatedLegacyControllerContext(function () use ($fusionPath) {
```
h

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -298,6 +264,43 @@ public function getLastEvaluationStatus()
+        return $this->withSimulatedLegacyControllerContext(function () use ($fusionPath) {
+            try {
+                $output = $this->render($fusionPath);
+            } finally {
+                $this->popContext();
+            }
```
add this
```
} catch (RuntimeException $exception) {
    throw $exception->getPrevious();
}
```

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -929,6 +932,87 @@ protected function throwExceptionForUnrenderablePathIfNeeded($fusionPath, $fusio
+            $httpResponse = $httpResponse->withStatus($this->legacyActionResponseForCurrentRendering->getStatusCode());
+        }
+
+        // reset for next render
+        $this->legacyActionResponseForCurrentRendering = null;
```
unsafe without `try` `finally`

mhsdesign committed:
TASK: Runtime fix `renderResponse` lock not being released

mhsdesign committed:
TASK: Runtime `renderResponse` unwrap `RuntimeException` itself

Previously the pattern was that the utmost caller of the runtime would unwrap the exception.
To simplify this, as the runtime now has a single entry point, we add this behaviour into the runtime.

mhsdesign committed:
TASK: Add Fusion RuntimeException::getWrappedException

... to centralise tricking phpstan

mhsdesign committed:
TASK: Remove manual http response parsing from FusionExceptionView

mhsdesign committed:
TASK: Revert Fusion `FusionView` HttpResponse support

This will be discussed separately and not part of the change of the Neos Node `FusionView`

mhsdesign committed:
TASK: Make Fusion `FusionView` independent of `ControllerContext`

mhsdesign committed:
!!! TASK: Introduce `LegacyFusionControllerContext` instead of using `ControllerContext`

`getArguments` removed without replacement.

mhsdesign committed:
TASK: Fix passing `tags` to plugin

mhsdesign commented:
I can confirm that a `Neos.Neos:Plugin` works as described here https://github.com/neos/Neos.Demo/pull/190

Though it needs an additional patch as plugins are currently broken on 9.0-dev anyways: https://github.com/neos/neos-development-collection/issues/4909

Things i tested in the controller:

- forward (content of other action)
- throwStatus (500 will be upmerged and plugin will show text inside: `500 Internal Server Error`
- redirect (uri will change)
- set arbitrary header `x-mhs: foobar`
- build uri to another action
- render fluid view
- render string

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -298,6 +261,45 @@ public function getLastEvaluationStatus()
+                is_string($output),
+                $output instanceof \Stringable => Utils::streamFor((string)$output),
+                $output === null, $output === false => Utils::streamFor(''),
+                default => throw new \RuntimeException(sprintf('Cannot render %s into http response body.', get_debug_type($output)), 1706454898)
```
if json serializeable set header and serialize

mhsdesign committed:
!!! BUGFIX: Fusions `FusionView` will return `ResponseInterface` instead of `mixed`

This will make the `Neos.Fusion:Http.Message` prototype work out of the box.
This was also done to return a value allowed by the view:
`string|ActionResponse|ResponseInterface|StreamInterface|\Stringable`

Adjusts AbstractFusionObjectTest::buildView to the runtime directly.
We still use a `ViewInterface` here as this allows us to not rewrite all our tests :D
We will rewrite them either way in behat some time.

mhsdesign committed:
TASK: `FusionExceptionView` handles psr response correctly

mhsdesign committed:
TASK: `Runtime::renderResponse` will try to jsonSerialize result if not a string

After a discussion with Christian we found it to be a more workable behaviour than throwing an exception.

mhsdesign commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -298,6 +260,55 @@ public function getLastEvaluationStatus()
+                return new Response(body: $output);
+            }
+
+            if (is_array($output) || $output instanceof \JsonSerializable || $output instanceof \stdClass || is_bool($output)) {
+                try {
+                    $jsonSerialized = json_encode($output, JSON_THROW_ON_ERROR);
+                } catch (\JsonException $e) {
+                    throw new \RuntimeException(sprintf('Cannot render %s into http response body.', get_debug_type($output)), 1708713158, $e);
+                }
+                $jsonResponse = new Response(body: $jsonSerialized);
+                return $jsonResponse->withHeader('Content-Type', 'application/json');
+            }
```
todo revert auto json serialisation of a fusion path if its not a stringable

mhsdesign replied:
@kitsunet fyi, i discussed this pr with Wilhelm and we concluded that this magic here will probably not serve very well.

The points were
- it is not expected if not looking into the code that special handling like this exists
  - i would not use it myself but rather be explicit as well
  - there is only a minimal gain for people knowing this feature and a http json response is more obvious and correct and does no harm
- currently the action controller will just silently ignore array values returned from the view, but this would change the behaviour and display accidentally things that are not noticed

thus we throw a nice exception:

> Fusion entry path "root" is expected to render a compatible http response body: string|\Stringable|null. Got array instead.

kitsunet replied:
Do we want to support streams here too maybe? Not sure if that can be useful in fusion but who knows... But yeah sounds good to me!

mhsdesign replied:
Streams are kindof impractical to create and against the fusion cache dna ... implicit support exist because streams are `\Stringable` and utils::streamFor will just return a stream if it finds one.

Still fusion would need a complete rewrite to really support streams

mhsdesign committed:
TASK: Improve `fusionGlobals` handling in the Fusion `FusionView`

Initially introduced the option expected an actual instance of `FusionGlobals`. This is a flawed design as only primitives can be set for example via `Views.yaml`

Additionally, the `FusionView` is forward compatible for this change https://github.com/neos/flow-development-collection/pull/3286
The `fusionGlobals` must not be used for setting the request but rather

```
$view->assign('request"', $this->request)
```

mhsdesign committed:
TASK: Runtime remove magic fusion jsonSerialize again

This magic behaviour is not expected and will rather lead to bugs instead of help the integrator to use fusion.

The action controllers `renderView` would previously just silently ignore this case and render a blank page.

Instead, to force to write correct functioning entry points, we throw an exception like:

> Fusion entry path "root" is expected to render a compatible http response body: string|\Stringable|null. Got array instead.

mhsdesign committed:
TASK: Refine fusion inline docs

mhsdesign commented:
Fyi the Fusion runtime will be officially marked internal with this change: https://github.com/neos/neos-development-collection/pull/4919 so we dont have to worry about polluting the api with this change as its _not_ api ^^ ;)

mhsdesign commented on Neos.Fusion/Classes/View/FusionView.php:
```diff
@@ -139,13 +175,16 @@ public function setFusionPathPatterns(array $pathPatterns)
      * @api
      */
-    public function render()
+    public function render(): ResponseInterface
```
We discussed that we want to return PsrStream | PsrResponse here, which will be enforced by https://github.com/neos/flow-development-collection/pull/3286

that way its less breaking for simple string renderings as the PsrStream is stringable.

mhsdesign committed:
TASK: Adjust views to stricter return types

mhsdesign committed:
TASK: Return `ResponseInterface` or `StreamInterface` in fusion view

- this will make the change less breaking due to allowing an implicit string cast
- adjust to fluidViews will now return a stream

mhsdesign committed:
TASK: Adjust to strict types in `ViewInterface`

grebaldi commented on Neos.Fusion/Classes/Core/LegacyFusionControllerContext.php:
```diff
@@ -0,0 +1,127 @@
+     *
+     * Gives access to the legacy mutable action response simulation {@see Runtime::withSimulatedLegacyControllerContext()}
+     *
+     * Initially it was possible to mutate the current response of the active MVC controller though this getter.
```
```suggestion
     * Initially it was possible to mutate the current response of the active MVC controller through this getter.
```

grebaldi commented on Neos.Fusion/Classes/Core/LegacyFusionControllerContext.php:
```diff
@@ -0,0 +1,127 @@
+        return $this->legacyActionResponseForCurrentRendering;
+    }
+
+    /**
+     * The method {@see ControllerContext::getArguments()} was removed without replacement.
+     */
+    // public function getArguments(): Arguments;
```
What's the reason for keeping it around?

mhsdesign replied:
it out-commented so in case someone looks up this code it will be clear that it was once part of the controller context but is not implemented here ;) Just a nice dev hint. I expect this class to be removed with Neos 10 or later so it doest have to be super pretty ^^

grebaldi commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -933,6 +951,78 @@ protected function throwExceptionForUnrenderablePathIfNeeded($fusionPath, $fusio
+    /**
+     * Implements the legacy controller context simulation {@see self::getControllerContext()}
+     *
+     * Initially it was possible to mutate the current response of the active MVC controller though $response.
```
```suggestion
     * Initially it was possible to mutate the current response of the active MVC controller through $response.
```

grebaldi commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -933,6 +951,78 @@ protected function throwExceptionForUnrenderablePathIfNeeded($fusionPath, $fusio
+     *
+     * Initially it was possible to mutate the current response of the active MVC controller though $response.
+     * While HIGHLY internal behaviour and ONLY to be used by Neos.Fusion.Form or Neos.Neos:Plugin
+     * this legacy layer is in place still allows this functionality.
```
```suggestion
     * this legacy layer is in place to still allow this functionality.
```

grebaldi commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -933,6 +951,78 @@ protected function throwExceptionForUnrenderablePathIfNeeded($fusionPath, $fusio
+
+        // actual rendering
+        try {
+            $httpResponseOrStream = $renderer();
```
```suggestion
        // actual rendering
        $httpResponseOrStream = null;
        try {
            $httpResponseOrStream = $renderer();
```

for clarity

grebaldi commented on Neos.Fusion/Classes/Core/LegacyFusionControllerContext.php:
```diff
@@ -0,0 +1,127 @@
+     * Initially it was possible to mutate the current response of the active MVC controller though this getter.
+     *
+     * While *HIGHLY* internal behaviour and *ONLY* to be used by Neos.Fusion.Form or Neos.Neos:Plugin
+     * this legacy layer is in place still allows this functionality.
```
```suggestion
     * this legacy layer is in place to still allow this functionality.
```

grebaldi commented on Neos.Fusion/Tests/Functional/FusionObjects/AbstractFusionObjectTest.php:
```diff
@@ -27,40 +28,72 @@
+     * TODO THIS IS HACKY AS WE CREATE AN OWN VIEW
      *
-     * @return FusionView
+     * We do that as the FusionView (rightfully) does return mixed anymore.
```
```suggestion
     * We do that as the FusionView (rightfully) doesn't return mixed anymore.
```

grebaldi commented on Neos.Fusion/Tests/Functional/View/FusionViewTest.php:
```diff
@@ -23,45 +25,71 @@
      * @test
      */
-    public function fusionViewUsesGivenPathIfSet()
+    public function fusionVieReturnsStreamInterface()
```
```suggestion
    public function fusionViewReturnsStreamInterface()
```

grebaldi reviewed (commented):
Hi @mhsdesign,

this is a tough one. Thanks so much for taking care of it! :100: I imagine it must have been quite a pain to track down all the implications of this API change (I for one am definitely struggling to grasp the full extent of it). So, kudos for your efforts!

I left some comments, but those are almost all typos in comments and test methods that stuck out to me. 

The overall idea of adding a prosthesis for `ControllerContext` to please the servants of dark magic is as good as it gets, I suppose. It sure ain't easy to get rid of that thing. :fire: 

I will do some testing with Fusion Form on top of this and get back to you with more feedback.

**EDIT:** Just tested it with Fusion Form, and it works perfectly :+1:

mhsdesign committed:
Fix typos

Co-authored-by: Wilhelm Behncke <2522299+grebaldi@users.noreply.github.com>

kitsunet commented on .github/workflows/build.yml:
```diff
@@ -132,6 +133,12 @@ jobs:
           git -C ../${{ env.NEOS_FOLDER }} checkout -b build
           composer config repositories.neos '{ "type": "path", "url": "../${{ env.NEOS_FOLDER }}", "options": { "symlink": false } }'
           composer require --no-update neos/neos-development-collection:"dev-build as ${{ env.NEOS_BRANCH_ALIAS }}"
+          # TODO REMOVE BEFORE MERGE!!!!!
```
Mmm, I guess this neesd to happen still?

mhsdesign replied:
Yes always these details and hacks used to fix ci for those complex things 🙈 ...

mhsdesign committed:
TASK: Introduce `TestingViewForFusionRuntime` for testing instead of using a hacky anonymous class

kitsunet reviewed (approved):
Yes!
